File: homework_4/matlab/part2/Test_deblur.m
% % % % % % % % % % % % % % % % % % % % % % % % % % % % % % % % % % % % % %
% % % EE556 - MATHMATICS OF DATA  % % % % % % % % % % % % FALL 2017 % % % %
% % % % % % % % % % % % % % % % % % % % % % % % % % % % % % % % % % % % % %
% BLIND DECONVOLUTION BY CONVEX PROGRAMMING % % % % % % % % % % % % % % % % 
% % % % % % % % % % % % % % % % % % % % % % % % % % % % % % % % % % % % % %
% PURPOSE: We will solve the following problem formulation with % % % % % %
% Frank-Wolfe's method.                                    
%                   min_x  0.5*norm(A(x) - b)^2        
%                   s.t.:  norm_nuc(x) <= kappa,     
%
% Laboratory for Information and Inference Systems (LIONS)
% Ecole Polytechnique Federale de Lausanne (EPFL) - SWITZERLAND
% Last modification: November 23, 2017
% % % % % % % % % % % % % % % % % % % % % % % % % % % % % % % % % % % % % %
% % % % % % % % % % % % % % % % % % % % % % % % % % % % % % % % % % % % % %

%% Clear varibles, add dependencies, etc.
clearvars;
addpath('PROPACK');

%% Mappings between n dimensional complex space and 2n dimensional real space
real2comp = @(x) x(1:size(x,1)/2) + 1i*x(size(x,1)/2+1:end);
comp2real = @(x) [real(x); imag(x)];

%% Load Data
x = double(rgb2gray(imread('blurredplate.jpg')));
x = x(61:188,41:296);
x = x./max(abs(x(:)));
figure(1)
imshow(x)
%x = x/norm(x,'fro');
imsize1 = size(x,1); 
imsize2 = size(x,2); 
imsize = imsize1*imsize2; 

%% Reshaping operators matrix to vector and vector to matrix
mat = @(x) reshape(x,imsize1,imsize2);
vec = @(x) x(:);

%% Set the measurement vector b
b = comp2real(vec(fft2(fftshift(x))));

%% Roughly estimate the support the blur kernel
K1 = 17;
K2 = 17;
Indw = zeros(imsize1,imsize2);
Indw(imsize1/2-(K1+1)/2+2:imsize1/2+(K1+1)/2,imsize2/2-(K2+1)/2+2:imsize2/2+(K2+1)/2) = 1;
% above, for simplicity we assume K1 and K2 odd,
% if they are not 1 pixel probably won't cause much trouble
figure(2)
imshow(reshape(ones - Indw,[imsize1, imsize2]))
title('estimated support of blur kernel');
Indw = Indw(:);
kernelsize = nnz(Indw);
[Indi,Indv] = find(Indw);

%% Define operators Bop and Cop
Bmat = sparse(Indi, 1:kernelsize, Indv, imsize, kernelsize );
Bop  = @(x) mat(Bmat*x);
BTop = @(x) Bmat'*vec(x);

%% Compute and display wavelet coefficients of the original and blurred image
[~,l] = wavedec2(x,4,'db1');
Cop  = @(x) waverec2(x,l,'db1');
CTop = @(x) wavedec2(x,4,'db1')';

%% Define operators
% We provide three operators:
% 1- Aoper(m,n,h) is equal to A*(m*n*h'). In other words, Aoper applies linear operator A
% to the rank 1 matrix (m*n*h'). Implmenting a special instance of A, we gain storage
% and computational efficiencies. Note that in Frank-Wolfe, we need operator A to 
% compute the gradient A'*(AX - b). Note that X in general is not rank 1, but we can 
% compute AX using Aoper because the updates on X are rank one. Using linearity, we can
% keep track of AX directly. See the function Frank-Wolfe for more details. 
% 2- AToper.matvec(y,w) computes (A'*y)*x, without generating A'*y in the ambient 
% dimensions. By doing this, we gain storage and computational efficiencies. Note that we
% need A' to compute the gradient A'*(AX-b). Be careful here, that you do not need to form
% the gradient A'*(AX-b), and in fact you should avoid it. We only need to run lmo with 
% the gradient. We can use the power method or the Lanczos method (eigs) to compute 
% top singular vectors. Both of these methods relies on multiplying A'*(AX-b) by a random 
% vector again and again. Hence, we only need an operator that computes (A'*(AX-b))*w for
% a given x. Hence, we actually use AToper.matvec(AX-b,w) to solve lmo. 
% 3- Similar to AToper.matvec(y,w), but this time we compute w'*(A'*y), and will be used 
% in the lmo. 
Aoper = @(m, n, h) comp2real(1/sqrt(imsize)*n*vec(fft2(Cop(m)).*fft2(Bop(h))));
AToper.matvec  = @(y, w) CTop(real(fft2(mat(conj(real2comp(y)).*vec(fft2(Bop(w)))))))/sqrt(size(y,1)/2);
AToper.rmatvec = @(y, w) BTop(real(ifft2(mat(real2comp(y).*vec(ifft2(Cop(w)))))))*(size(y,1)/2)^1.5;

%% Run Frank-Wolfe's method
MaxIters = 200;
kappa = 1000;
[xFW, outFw] = FrankWolfe( Aoper, AToper, b, kernelsize, imsize, kappa, MaxIters, @(m)plotFunc(m,Cop,x));

%% NOTE: This experiment is based on the theory and the codes publised in
% 'Blind Deconvolution using Convex Programming' by A.Ahmed, B.Recht and
% J.Romberg.

File: homework_4/python/part2/Test_deblur.py

# coding: utf-8

# In[ ]:

import numpy as np
from scipy.fftpack import fft2, fftshift, ifft2
import matplotlib.pyplot as plt
import scipy.sparse as sps
from scipy.misc import imread
from scipy.sparse.linalg import LinearOperator, svds
from pywt import wavedec2, waverec2, coeffs_to_array, array_to_coeffs
from collections import namedtuple
from math import sqrt

import time
import pylab as pl
from IPython import display


# In[ ]:

# Mappings between n dimensional complex space and 2n dimensional real space
real2comp = lambda x: x[0:x.shape[0]/2] + 1j*x[x.shape[0]/2:]
comp2real = lambda x: np.append(x.real, x.imag)

# Load Data
x = imread('blurredplate.jpg',flatten=True,mode='F')
x = x[60:188,40:296]
x = x/np.linalg.norm(x,ord=np.inf)
#x = x/np.linalg.norm(x)
imsize1 = x.shape[0]
imsize2 = x.shape[1]
imsize = imsize1*imsize2

# Shows the blurred image!
plt.imshow(x, cmap='gray')
plt.show()

# Reshaping operators matrix to vector and vector to matrix
mat = lambda x: np.reshape(x,[imsize1,imsize2])
vec = lambda x: x.flatten()

# Set the measurement vector b
b = comp2real(vec(fft2(fftshift(x))))

# Roughly estimate the support of the blur kernel
K1 = 17
K2 = 17
Indw = np.zeros([imsize1,imsize2])
ind1 = np.int(imsize1/2-(K1+1)/2+1)
ind2 = np.int(imsize1/2+(K1+1)/2)
ind3 = np.int(imsize2/2-(K2+1)/2+1)
ind4 = np.int(imsize2/2+(K2+1)/2)
Indw[ind1:ind2,ind3:ind4] = 1
#above, for implementational simplicity we assume K1 and K2 odd, even  
#if they are even 1 pixel probably won't cause much trouble
plt.imshow(-Indw, cmap='gray') # Shows the estimated support of blur kernel!
plt.show()
Indw = vec(Indw);
kernelsize = np.count_nonzero(Indw)
Indi = np.nonzero(Indw > 0)[0]
Indv = Indw[Indi]

# Define operators Bop and Cop
Bmat = sps.csr_matrix((Indv,(Indi,range(0,kernelsize))),shape=(imsize,kernelsize))
Bop = lambda x: mat(Bmat.dot(x))
BTop = lambda x: Bmat.T.dot(vec(x))

# Compute and display wavelet coefficients of the original and blurred image
l = coeffs_to_array(wavedec2(x, 'db1', level=4))[1]
Cop = lambda x: waverec2(array_to_coeffs(mat(x),l,output_format='wavedec2'),'db1')
CTop = lambda x: coeffs_to_array(wavedec2(x, 'db1', level=4))[0]

# Define operators
Aoper = lambda m,n,h: comp2real(1.0/sqrt(imsize)*n*vec(fft2(Cop(m))*fft2(Bop(h))))
AToper = {"matvec": lambda y,w: CTop(np.real(fft2(mat(np.conj(real2comp(y))*vec(fft2(Bop(w)))))))/sqrt(y.shape[0]/2.0),
         "rmatvec": lambda y,w: BTop(np.real(ifft2(mat(real2comp(y)*vec(ifft2(Cop(w)))))))*(y.shape[0]/2.0)**1.5}


# In[ ]:

def plotFunc(mEst, C, x):
    
    xEst = -C(mEst);
    xEst = xEst - min(xEst.flatten())
    xEst = xEst/max(xEst.flatten())
    plt.imshow(xEst, cmap='gray',hold=False)
    plt.show()


# In[ ]:

def FrankWolfe(Aoper, AToper, b, n1, n2, kappa, maxit, plotFunc):
# PURPOSE: We will solve the following problem formulation with
# Frank-Wolfe's method.                                    
#                   min_x  0.5*norm(A(x) - b)^2        
#                   s.t.:  norm_nuc(x) <= kappa,     
#
# Laboratory for Information and Inference Systems (LIONS)
# Ecole Polytechnique Federale de Lausanne (EPFL) - SWITZERLAND
# Last modification: November 26, 2017
    
    # Print the caption
    
    #Initialize 
    AX_t = 0.0   # zeros
    X = 0.0      # zeros
    
    # keep track of objective value
    fx = np.array([])
        
    # The main loop    
    for iteration in range(0, maxit+1):
        
        # Print the objective values ...
        fx = np.append(fx, 0.5*np.linalg.norm(AX_t - b,2)**2)
        print('{:03d} | {:.4e}'.format(iteration, fx[-1]))

        # Form the residual and fix the operator to be used in svds.
        res_cur = AX_t - b

        ATop1 = lambda w: AToper["matvec"](res_cur,w)
        ATop2 = lambda w: AToper["rmatvec"](res_cur,w)
        svdsArg = LinearOperator((n2,n1), matvec=ATop1, rmatvec=ATop2)
        topLe_vec, singVal, topRe_vec = svds(svdsArg, k=1, tol=1e-4, which='LM')
        # Note: we could also used svds. Lansvd and svds solve the same problem with similar
        # but different approaches. Svds in older versions of Matlab does not accept function
        # handles as inputs, this is why we rather used lansvd here. If you run into trouble
        # with lansvd on your computer, try to use svds (with properly modifying the inputs)
        
        # Apply A to the rank 1 update
        AXsharp_t = Aoper(topLe_vec, -kappa, topRe_vec.T)
        
        # Step size
        weight = 2.0/(iteration+2)
        
        # Update A*X
        AX_t = (1.0-weight)*AX_t + weight*(AXsharp_t)
        
        # Update X
        X = (1.0 - weight)*X + weight*(-kappa*np.outer(topLe_vec, topRe_vec))
        
        # Show the reconstruction (at every 10 iteration) 
        if (iteration%200 == 0):
            U,S,V = np.linalg.svd(X,full_matrices=0,compute_uv=1)
            plotFunc(U[:,0])

    return X

# In[ ]:

# Run Frank-Wolfe's method
MaxIters = 200
kappa = 1000
plotF = lambda m: plotFunc(m,Cop,x)
xFW = FrankWolfe(Aoper, AToper, b, kernelsize, imsize, kappa, MaxIters, plotF)


# In[ ]:

# NOTE: This experiment is based on the theory and the codes publised in
#'Blind Deconvolution using Convex Programming' by A.Ahmed, B.Recht and J.Romberg.


# In[ ]:



